### PR TITLE
enhancement: introduce Ruff for Python linter for reordering and removing unused imports with automated pre-commit and sytle check

### PR DESCRIPTION
**Pull request by bowenliang123**
- The Ruff tool is a general linting and formatting tools for Python code
- Ruff is blazingly fast, as it finished lining or fixing all the hundreds of Python scripts of Dify project in less than a second, with the support of local cache and the core written in Rust
- Ruff linter is designed as drop-in replacement for [Flake8](https://pypi.org/project/flake8/) (plus dozens of plugins), [isort](https://pypi.org/project/isort/), [pydocstyle](https://pypi.org/project/pydocstyle/), [pyupgrade](https://pypi.org/project/pyupgrade/), [autoflake](https://pypi.org/project/autoflake/), and more
- To sort Python imports, the isort rules (I001, I002) in Ruff are applied (https://docs.astral.sh/ruff/rules/#isort-i)
   - although the line breaks in a style closer to `black` 
- Re-arranged Python imports in sections separated in source parties for better readability, in the order of "future", "standard-library", "third-party", "first-party", "local-folder". This greatly improves the readability and understandability.
- To remove the unused imports, F401 rule of Flake8 in Ruff is applied (https://docs.astral.sh/ruff/rules/#pyflakes-f)
   - fixed 190 violations
   - the tests directory is skipped as the tests requires some unused-like imports for importing mocking methods

The key changes in this PR:
1. apply ruff auto-fixes in one-key linting script `dev/reformat`
2. automatically run linter scripts in `pre-commit` (via husky support) when committing changes, if API module changes
3. check Python style in CI style workflow by running ruff linter and list all the violations if it fails

### CODE REVIEW

File: dev/reformat [hunk 1]
```diff
@@ -2,10 +2,11 @@
...
-else
-    isort  --settings ./.github/linters/.isort.cfg ./
+# python style checks rely on `ruff` in path
+if ! command -v ruff &> /dev/null; then
+    echo "Installing Ruff ..."
+    pip install ruff
 fi
+
+# run ruff linter
+ruff check --fix ./api
```
Comment (bowenliang123) [inline dev/reformat:12]: 1. apply ruff auto-fixes in one-key linting script

File: web/.husky/pre-commit [hunk 2]
```diff
@@ -1,4 +1,35 @@
...
+        web_modified=true
+    fi
+done
+
+# run linters based on the modified modules
+
+if $api_modified; then
+    echo "Running Ruff linter on api module"
+    ./dev/reformat
+fi
```
Comment (bowenliang123) [inline web/.husky/pre-commit:28]: 2. automatically run linter scripts in pre-commit when committing changes, if API module changes

File: .github/workflows/style.yml [hunk 3]
```diff
@@ -18,6 +18,16 @@ jobs:
...
 
+      - name: Set up Python
+        uses: actions/setup-python@v5
+        with:
+          python-version: '3.10'
+
+      - name: Python Style Check
+        run: |
+          pip install ruff
+          ruff check ./api
```
Comment (bowenliang123) [inline .github/workflows/style.yml]: 3. check Python style by running ruff linter and list all the violations if it fails

**Review by bowenliang123 — COMMENTED** [review]
Key points of this PR.

File: api/pyproject.toml [hunk 4]
```diff
@@ -0,0 +1,17 @@
...
+]
+line-length = 120
+
+[tool.ruff.lint]
+ignore-init-module-imports = true
+select = [
+    "F401", # unused-import
+    "I001", # unsorted-imports
+    "I002", # missing-required-import
+]
```
Comment (bowenliang123) [inline api/pyproject.toml:17]: Three rules are enabled by demand. And the list is extendable for other linter rules of Flake8, pycodestyle and etc.   (refer to https://docs.astral.sh/ruff/rules)

**Comment by bowenliang123** [comment]
cc @takatost

File: web/.husky/pre-commit [hunk 5]
```diff
@@ -1,4 +1,35 @@
-#!/usr/bin/env sh
+#!/usr/bin/env bash
 . "$(dirname -- "$0")/_/husky.sh"
```
Comment (takatost) [inline web/.husky/pre-commit:2]: Maybe this should be moved to the block of `web_modified`?

**Review by takatost — CHANGES_REQUESTED** [review]
I have a question. If the developer didn't initialize the pre-commit hook in the `web` directory, does that mean the Python reformat pre-commit hook won't work?

**Comment by bowenliang123** [comment]
> I have a question. If the developer didn't initialize the pre-commit hook in the `web` directory, does that mean the Python reformat pre-commit hook won't work?

Yes. The workaround is that when the Python style CI job fails, it prints instructions to suggest the user to run `dev/reformat` which is clear and easy to apply.

As for the usage of Husky and other practices, they should be done in separated PR. As for this time, it's preferred to reuse the current Husky tool's support. The Git hooks require commit a `pre-commit` in `.git/hooks/pre-commit` which is ignored by default in `.git`, so I think we still need the necessary tool to automate the injection to the target script destination, with Husky or pre-commit.com tool or sth else.

File: web/.husky/pre-commit (hunk 5 above)
Comment (bowenliang123) [reply web/.husky/pre-commit:2]: Seems it's part of Husky required for Windows. I would prefer not to change it for this time.
https://typicode.github.io/husky/troubleshoot.html#yarn-on-windows

**Review by crazywoola — DISMISSED** [review]
LGTM @iamjoel Please take a look at this pr as well.

**Comment by bowenliang123** [comment]
Rebased and re-applied the Ruff fixes on to the latest commit of the master branch to resolve conflicts.

**Review by crazywoola — CHANGES_REQUESTED** [review]
This is a new conflict in files.

**Comment by bowenliang123** [comment]
> This is a new conflict in files.

Resolved.

**Review by crazywoola — APPROVED** [review]

**Review by takatost — APPROVED** [review]

**Pull request merged**